client/layout/masterbar-item-notifications.jsx

```diff
@@ -0,0 +1,140 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import ReactDom from 'react-dom';
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import MasterbarItem from './masterbar-item';
+import Notifications from 'notifications';
+import store from 'store';
+
+export default React.createClass( {
+	displayName: 'MasterbarItemNotifications',
+
+	propTypes: {
+		user: React.PropTypes.object,
+		isActive: React.PropTypes.bool,
+		className: React.PropTypes.string,
+		onClick: React.PropTypes.func,
+	},
+
+	getInitialState() {
+		let newNote = false;
+		let user;
+
+		if ( this.props.user ) {
+			user = this.props.user.get();
+		}
+
+		// User object should be loaded by now, but
+		// if it isn't just wait until the notifications
+		// finish their initial load to set `newNote`
+		if ( user && user.has_unseen_notes ) {
+			newNote = true;
+		}
+
+		return {
+			isShowingPopover: false,
+			newNote: newNote,
+			animationState: 0,
+		};
+	},
+
+	checkToggleNotes( event, forceToggle ) {
+		const target = event ? event.target : false;
+		const notificationNode = this.getNotificationLinkDomNode();
+
+		if ( notificationNode.contains( target ) ) {
+			return;
+		}
+
+		if ( this.state.isShowingPopover || forceToggle === true ) {
+			this.toggleNotesFrame( event );
+		}
+	},
+
+	toggleNotesFrame( event ) {
+		if ( event ) {
+			event.preventDefault();
+		}
+
+		this.setState( {
+			isShowingPopover: ! this.state.isShowingPopover
+		}, () => {
+			this.props.onClick( this.state.isShowingPopover );
+
+			if ( this.state.isShowingPopover ) {
+				this.setNotesIndicator( 0 );
+			}
+
+			// focus on main window if we just closed the notes panel
+			if ( ! this.state.isShowingPopover ) {
+				this.getNotificationLinkDomNode().blur();
+				window.focus();
+			}
+		} );
+	},
+
+	getNotificationLinkDomNode() {
+		return ReactDom.findDOMNode( this.refs.notificationLink );
+	},
+
+	/**
+	 * Uses the passed number of unseen notifications
+	 * and the locally-stored cache of that value to
+	 * determine what state the notifications indicator
+	 * should be in: on, off, or animate-to-on
+	 *
+	 * @param {Number} currentUnseenCount Number of reported unseen notifications
+	 */
+	setNotesIndicator( currentUnseenCount ) {
+		let existingUnseenCount = store.get( 'wpnotes_unseen_count' );
+		let newAnimationState = this.state.animationState;
+
+		// Having no record of previously unseen notes is
+		// functionally equal to having a record of none
+		if ( null === existingUnseenCount ) {
```

Comment:
This may be a good time to get rid of this block. `null` converts to 0 for comparisons, so it is redundant.